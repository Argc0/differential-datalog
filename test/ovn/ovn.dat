start;

insert OVN_Northbound.Logical_Switch (
    ._uuid = 0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d95,
    .name = "lsw0",
    .ports = [0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d96, 0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d97],
    .acls = [0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d98, 0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d99],
    .qos_rules = [0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d9a, 0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d9b],
    .load_balancer = [0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d9c, 0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d9d],
    .dns_records = [0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d9e, 0x06a2cf7c_7ea1_4ca2_a0a2_6176fd1c6d9f],
    .other_config = [("c1", "v1"), ("c2", "v2")],
    .external_ids= []
);
commit;
dump;

start;
insert HostAddresses(0, ["10.10.10.101", "10.10.10.102", "10.10.10.103"]),
insert HostAddresses(0, ["10.10.10.103", "10.10.10.104", "10.10.10.105"]),
insert HostAddresses(1, ["192.168.0.1", "192.168.0.2", "192.168.0.3"]);
commit;

dump HostAddress;


start;

insert ExternalIds(0, [("a", "1"),("b", "2"),("c", "3")]),
insert ExternalIds(0, [("c", "4"),("d", "5"),("e", "6")]);
commit;

dump ExternalId;
