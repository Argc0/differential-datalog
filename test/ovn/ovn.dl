import OVN_Northbound
import OVN_Southbound


// FlatMap test
input relation HostAddresses(host: bit<64>, addrs: Set<string>)
relation HostAddress(host: bit<64>, addr: string)

HostAddress(host, addr) :- HostAddresses(host, addrs),
                           FlatMap(addr=addrs).


input relation ExternalIds(host: bit<64>, ids: Map<string, string>)
relation ExternalId(host: bit<64>, id: (string, string))

ExternalId(host, id) :- ExternalIds(host, ids),
                        FlatMap(id=ids).
