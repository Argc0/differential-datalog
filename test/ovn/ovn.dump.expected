OVN_Northbound_Logical_Switch:
OVN_Northbound.Logical_Switch{8820728388062355172047994593891610005,"lsw0",std_Set { x: {8820728388062355172047994593891610006, 8820728388062355172047994593891610007} },std_Set { x: {8820728388062355172047994593891610008, 8820728388062355172047994593891610009} },std_Set { x: {8820728388062355172047994593891610010, 8820728388062355172047994593891610011} },std_Set { x: {8820728388062355172047994593891610012, 8820728388062355172047994593891610013} },std_Set { x: {8820728388062355172047994593891610014, 8820728388062355172047994593891610015} },std_Map { x: {"c1": "v1", "c2": "v2"} },std_Map { x: {} }}

HostAddress{0,"10.10.10.101"}
HostAddress{0,"10.10.10.102"}
HostAddress{0,"10.10.10.103"}
HostAddress{0,"10.10.10.104"}
HostAddress{0,"10.10.10.105"}
HostAddress{1,"192.168.0.1"}
HostAddress{1,"192.168.0.2"}
HostAddress{1,"192.168.0.3"}
ExternalId{0,("a", "1")}
ExternalId{0,("b", "2")}
ExternalId{0,("c", "3")}
ExternalId{0,("c", "4")}
ExternalId{0,("d", "5")}
ExternalId{0,("e", "6")}
